### PR TITLE
qt5: 5.12.7 -> 5.14.2

### LINKED ISSUE #74355: qt 5.13.x
qt 5.13 has been released for a while now.

### LINKED ISSUE #81894: Update shiboken (eventually)
@gebner pending an update of Qt to > 5.12.3, https://github.com/NixOS/nixpkgs/blob/0107ee8c322e57cdb2ebcc1c9c4286ff7db53d5c/pkgs/development/python-modules/shiboken2/nix_compile_cflags.patch does not apply cleanly to newer versions of shiboken.

re: https://github.com/NixOS/nixpkgs/issues/74355

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change
Fixes #74355
Fixes #81894

###### To-Do list:

- [x] Fix loading of shared libraries as observed in Plasma and Quassel Core (fixed by @DizFer, thanks!)
- [ ] Update the following broken packages or move them back to Qt 5.12 if no fix is available
  - [x] boomerang
  - [x] chatterino2
  - [ ] deepin.* (does not build on master)
  - [x] diffpdf
  - [x] digikam
  - [x] falkon
  - [x] golden-cheetah
  - [x] ibus-engines.mozc
  - [x] lumina.lumina-pdf
  - [x] megasync
  - [x] lxqt.*
  - [x] mpc-qt
  - [ ] netease-cloud-music (does not start on master)
  - [x] nheko
  - [x] notepadqq
  - [x] qgroundcontrol
  - [x] qlcplus
  - [x] qtox
  - [x] quassel
  - [x] quaternion
  - [x] rockbox_utility
  - [x] rssguard
  - [x] sdrangel
  - [x] smplayer
  - [x] smtube

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/qt-5/modules/qtmultimedia.nix
Comment (cole-h) [inline pkgs/development/libraries/qt-5/modules/qtmultimedia.nix:13]:
```suggestion
    ++ optional (versionAtLeast qtbase.version "5.14.0" && stdenv.isLinux) wayland;
```

I think this should clear up the eval error.

Comment (petabyteboy) [reply pkgs/development/libraries/qt-5/modules/qtmultimedia.nix]: Thanks!

**Comment by petabyteboy** [comment]
qt5.qttools fails to build, becaus:

`Error copying /build/qttools-everywhere-src-5.14.2/lib/cmake/Qt5AttributionsScannerTools/Qt5AttributionsScannerToolsConfigVersion.cmake to /nix/store/v52ixjwqg4dqa65qcqfrhd9b9fp8za3a-qttools-5.14.2/lib/cmake/Qt5AttributionsScannerTools/Qt5AttributionsScannerToolsConfigVersion.cmake: Cannot open /build/qttools-everywhere-src-5.14.2/lib/cmake/Qt5AttributionsScannerTools/Qt5AttributionsScannerToolsConfigVersion.cmake for input`

But desktop applications like tdesktop and qutebrowser are working fine.

File: pkgs/development/libraries/qt-5/5.14/default.nix
Comment (cole-h) [inline pkgs/development/libraries/qt-5/5.14/default.nix:44]: Appears to redirect to https://github.com/qtwebkit/qtwebkit now. Maybe this should be updated?

File: pkgs/top-level/all-packages.nix
Comment (cole-h) [inline pkgs/top-level/all-packages.nix:14073]: Has this been addressed?

**Review by cole-h — COMMENTED** [review]
Diff LGTM, except for minor nits. No expert in this domain, so I'll defer to someone else (no way will my computer handle building and verifying the ~6000 affected packages lol).

**Comment by petabyteboy** [comment]
I will run nix-review at some point.
And good news: I was able to fix the qttools build! Will try to rebuild my system with this now.

File: pkgs/top-level/all-packages.nix
Comment (petabyteboy) [reply pkgs/top-level/all-packages.nix:14069]: I don't know where the qt version for darwin is set, and I don't have a darwin machine to test this on.

File: pkgs/development/libraries/qt-5/5.14/default.nix
Comment (petabyteboy) [reply pkgs/development/libraries/qt-5/5.14/default.nix:44]: Thanks, I will change it. This is exactly why it's good that you had a look even though you don't have specific knowledge about the topic :slightly_smiling_face:

**Comment by petabyteboy** [comment]
Now using this on my everyday system \o/
I had to override quassel src to master because it has some fixes to build with the new Qt. Otherwise it built fine after I fixed the QtAttributionsScanner problem in qttools.

**Comment by petabyteboy** [comment]
I've been working on the Plasma 5 NixOS test. I had to update KDE frameworks, plasma5, quassel and grandlee.

**Comment by petabyteboy** [comment]
I am unsure what is the best approach to get this new Qt version to NixOS users.

One possible approach is to set the new version as default from the beginning and try to fix as much as possible rightaway. This will probably be the quickest way to get as many applications as possible on the new Qt version, which is by the way required for some of the fixes to work (wayland clipboard will only be better in the applications that are built with the new Qt).
The downside of this will be the enourmous diff size and time required to review it. It might also lead to more user issues in the short-term, because I can't possibly test the functionality of all applications that use Qt.

Another approach would be to leave 5.12 as the default version and gradually move packages to the new version. This has the advantage that we can check every package thoroughly, but we risk that many applications will still use 5.12 in NixOS 20.09.

File: pkgs/development/libraries/qt-5/5.14/default.nix
Comment (petabyteboy) [reply pkgs/development/libraries/qt-5/5.14/default.nix]: According to the repository descriptions https://code.qt.io/cgit/qt/qtwebkit.git/ contains all snapshots, while the repository you mentioned also contains a lot of things that are only necessary for development. I changed it to https://github.com/qt/qtwebkit/ now, because this mirror offers a tarball.

File: pkgs/top-level/all-packages.nix
Comment (petabyteboy) [reply pkgs/top-level/all-packages.nix:14016]: Since we don't have any older version of Qt packaged in nixpkgs anymore I assume this was already done in a previous change. I will remove the comment, but it would still be great if someone could test the new version on darwin.

**Comment by ttuegel** [comment]
@petabyteboy In the past, we have done Qt 5 upgrades by changing the default version (the `qt5` attribute) and downgrading the packages that break. We will ping the maintainers of any downgraded packages, too. This is much less work than fixing all the packages, but gives a better result than trying to gradually upgrade (which won't happen). When Qt 5.12 is eventually EOL'd upstream, we will remove it and mark any packages still using it as `broken`.

**Comment by DIzFer** [comment]
Quassel situation isn't very nice: both core and client seem to work when updated to latest git commit, but that's _technically_ not a stable release, so probably wiser to kick it back to Qt5.12 until next release? Not sure how we would be supposed to remember to pull it back to the future once it's officially updated...

Other than that, I'm here because of qutebrowser, but something about QtWebEngine appears not to agree with my machine, so that I can't test. My desktop is at the moment rebuilding without QtWebEngine, which means I'll accidentally test some of https://github.com/NixOS/nixpkgs/pull/84542 too.

**Comment by petabyteboy** [comment]
Hey, let me see what I can do for you in terms of QtWebEngine and qutebrowser. I will set up a small repository that includes an overlay and a fixed version of nixpkgs, similar to colemickens' [nixpkgs-chromium](https://github.com/colemickens/nixpkgs-chromium) and push the build results to a Cachix cache.

About quassel: The daemon has problems finding it's database library, so even then it doesn't work. I'm just using the client with the latest master commit, but that shouldn't be something we commit to nixpkgs. Instead we should either backport the necessary patches to the last Quassel release or leave Quassel with Qt5.12 until there is a new release.

**Comment by DIzFer** [comment]
That would be most gracious of you, but nevermind, I've somehow successfully built qutebrowser and webengine without realising... As you can tell from my comment, nix frequently conspires against my expectations. Which in the end means this seems to build for me all right!

Besides a few other apps you don't mention to be broken not being broken, I checked on some others that might affect me:

 - nheko is supposedly fixed in https://github.com/NixOS/nixpkgs/pull/85922 (merged). Not been able to verify yet, since I'm apparently terrible at github-git translation.
- quaternion gets an update in https://github.com/NixOS/nixpkgs/pull/84160 (merged) which lets it both build and run.
 - digikam fails way before any real problem, an apparent checksum mismatch on https://github.com/NixOS/nixpkgs/pull/84232/commits/97ce35ea1637210081c9353a81e51b92eb922ca2#diff-c7c1cea8cd1fc212a3af4b8c39e28e87R47. Apparently we're six months behind, which may be due to KDE not syncing their [upstream repo](https://invent.kde.org/kde/digikam) to the [github mirror](https://github.com/kde/digikam) we download from. Unsure on how to proceed in this case (Update from upstream directly? Find whoever's responsible for KDE's mirrors on github so they can check what's up and wait until that's cleared up?).

I'll report with more details about digikam once I finish shaving the required yaks, which don't appear to be few...

**Comment by DIzFer** [comment]
Apparently a big enough yak and many yaks in a row look very similar. Digikam failed because building libqtav fails. libqtav has a new version github is displaying slightly different, apparently tags != releases? Whatever, but maybe that's why we haven't updated it automatically. In any case, latest libqtav doesn't build either, but HEAD felt like working. Currently 60% along building qtwebkit (_dear god why help me not another html engine_), and only one more derivation left before digikam's turn.

Additionally, reading through the list, rssguard caught my attention, I read up on it, and out of curiosity I decided to attempt to fix it. Once again, it's one release behind, but latest stable doesn't build either. Besides qmake being bollocks, I don't understand the error in question (`make[1]: *** No rule to make target '../../localization/qtbase_uk.qm', needed by 'rcc/qrc_rssguard.o'.  Stop.`), so I'm stopping here with this one.

**Comment by petabyteboy** [comment]
Thanks for helping with testing. I have rebased on top of master and confirmed the ones you mentioned to be working.

Unfortunately I ran into another issue that is giving me headaches:
When I first tested this Qt update I wasn't aware of the seperate update schedules for Plasma, kde-frameworks and Qt. Because of this and a build failure with Qt 5.14.2, kdeFrameworks 5.56 and Plasma 5.17.5, I tested the new Plasma 5.18.x and Qt 5.14.x in one branch. The Plasma settings application and some other things were not properly working on this branch, but back then I thought this issue was caused by the Plasma update, _which it is not_. The new Plasma with the old Qt works fine, while the old Plasma with the new Qt also has the issue.
**TL;DR:** Something with Qt applications loading shared libraries is seriously broken. I have observed this in the Plasma settings application and in Quassel core not being able to find its Postgres database adapter library.

 I had first thought to be related to the Plasma update. 

If anyone is in need of prebuilt QtWebEngine binaries for Qt5.14, maybe because they want to use a QtWebEngine-based browser with TLSv1.3 support, this is available [here](https://github.com/petabyteboy/nixpkgs-qutebrowser).

**Comment by DIzFer** [comment]
Right, the plasma issues make sense. You just prevented me from flooding the plasma update PR with panic and "everything is broken" comments :'D

Short version is, yeah, randomly missing QML modules, no wallpapers, most systemsettings KCMs are broken.

I just contacted the email listed at the github.com/kde organisation, in case that reaches someone that can fix that one.

So you go to all this effort to get a more secure QtWebEngine... and I just wanted to have a browser _not_ blind me in every website. Nice priorities, go me :P

**Comment by DIzFer** [comment]
I have news about digikam: It failed to build! Hooray!

When trying to re-build it with useful logs (`nix log` appears to be unhelpful this time, possibly because it failed before reaching buildPhase?), well, I was digging out how to update kde-frameworks, so they were now pointing to 5.69 (unstable has 5.68)... so I'm rebuilding the universe once again. [1/95/112] ATM, apparently everything remaining needs kio to be done.


While I'm at it, I'll see if the rest of plasma gets any better with new frameworks. Applications seems to be up-to-date.

**Comment by DIzFer** [comment]
> Something with Qt applications loading shared libraries is seriously broken. I have observed this in the Plasma settings application and in Quassel core not being able to find its Postgres database adapter library.

Wait, are Qt applications failing in general? I've only noticed brokenness with KDE stuffs, which would point to "updating frameworks" being a helpful step. But you're right, quassel doesn't seem to depend on that. Welp :(

**Comment by petabyteboy** [comment]
If you experience the "everything is broken" with the current state of the Plasma update PR, without the Qt update, please flood it with comments. I would expect most these problems to only occur when the Qt update is applied as well.

Multiple failures I have anlayzed were caused by an issue with the outputs of qttools. For example the mpc-qt build fails, because `qttools-5.14.2-bin/bin/lupdate` is called, but this does not exist. diffpdf fails becuase it tries to call `qttools-5.14.2-bin/bin/lrelease`, but this does not exist. Both binaries exist in the dev output and did not exist in the bin output of qttools in 5.12.

About the general problems with some applications built agains new Qt: Quassel does actually use quite a few of the k* libraries. I'm not sure if the problem lies in the kdeFrameworks, I haven't seen an application that is broken and doesn't use the k* libraries. I guess updating the frameworks would be worth a try, but I wouldn't count on it fixing the issue.

**Comment by DIzFer** [comment]
Progress with digikam: cmake failure, with viable logs. Seems like multiple outputs are messing up cmake, and the .cmake in exiv2-dev points to a file found in main exiv2. I know this must be easily fixable, but not how exactly. I'm parking this for now

> If you experience the "everything is broken" with the current state of the Plasma update PR, without the Qt update, please flood it with comments. I would expect most these problems to only occur when the Qt update is applied as well.

Yeah, between my tendency to try to do a hundred things at once and the resulting confusion, I'd expect at least some of my issues to be my own fault. I'll try to separate, but at some point these are going to have to learn to live together :P

For now, I confirm Qt 5.14 + Plasma 5.18 + Frameworks 5.69 fails miserably :(

Currently rebuilding Plasma 5.18 without Qt or Frameworks updates.

---

`nix show-derivation nixpkgs.quasselDaemon` doesn't seem to list anything from frameworks. The client does, though, so I'd expect that to be the reason frameworks are listed in the package .nix.

> Multiple failures I have anlayzed were caused by an issue with the outputs of qttools. For example the mpc-qt build fails, because qttools-5.14.2-bin/bin/lupdate is called, but this does not exist. diffpdf fails becuase it tries to call qttools-5.14.2-bin/bin/lrelease, but this does not exist. Both binaries exist in the dev output and did not exist in the bin output of qttools in 5.12.

The split in qttools looks fun. -bin has lupdate-pro, but lupdate lives in -dev. -dev also has a lupdate-pro, but it actually is a symlink to -bin. Many of the files in qttools-dev/bin are like that, a symlink to -bin/bin. Hopefully this make more sense if one understands what qttools is, but I don't.

**Comment by DIzFer** [comment]
Current progress with plasma: it appears the plasma brokenness when used with Qt 5.14 is because QML2_IMPORT_PATH isn't getting populated enough. For example, and most relevant to my day-to-day, Krunner doesn't run because it can't find the QML module for org.kde.milou.

> `infinite recursion encountered`

It appears milou currently depends on krunner, instead of the other way around...? Of course, because milou doesn't build, and I'm sure if I looked, the reasoning for the frameworks-applications-plasma split justifies this chain. Adding `milou` to plasma-workspace's buildInputs fixes this one in particular... I don't think it's reasonable to do this for everything just yet.

Sure enough, with what currently is nixos-unstable, the krunner wrapper does contain references to the milou derivation. Once you add your commit (93de786ea38ef6aebab56cd91705e30c8360a5d3), the new wrapper doesn't tell it how to find milou. Diff, after swapping nix store hashes with hash and qt versions with qtversion:

```diff
61,62d60
< export QT_PLUGIN_PATH='/nix/store/hash-milou-5.17.5/lib/qt-qtversion/plugins'${QT_PLUGIN_PATH:+':'}$QT_PLUGIN_PATH
< export QML2_IMPORT_PATH='/nix/store/hash-milou-5.17.5/lib/qt-qtversion/qml'${QML2_IMPORT_PATH:+':'}$QML2_IMPORT_PATH
155d152
< export XDG_DATA_DIRS='/nix/store/hash-milou-5.17.5/share'${XDG_DATA_DIRS:+':'}$XDG_DATA_DIRS
```

Alltogether, after some digging, it seems the logic in wrap-qt-apps-hook and company isn't working for Qt 5.14, so that's where I'm looking right now. I'm also hoping if this works for krunner, it will work for Quassel too :D

**Comment by DIzFer** [comment]
!

from https://wiki.qt.io/New_Features_in_Qt_5.14 :
 - New Features (within existing modules)
  - Qt Core
    - [QTBUG-14150](https://bugreports.qt.io/browse/QTBUG-14150) Qt is now relocatable, i.e. it's possible to move the Qt installation without breaking functionality or loading of plugins.

So, previous behaviour perfectly matched what Nix produces when building Qt: at build time, absolute paths to libraries are determined. Now, however, if conditions are "favorable", Qt defaults to just remembering where libQt5Core.so is, and tries to find the rest relative to it. Which doesn't work, since we split qtbase in what, in my ignorance, seems like a pretty weird way.

qtbase.bin contains lib/qt-qtversion/plugins/*, which includes, for example, the postgres plugin in sqldrivers/libqsqlpsql.so. However, qtbase.out is where lib/libQt5Core.so.5 lives.

What's probably happening, before stracing it (which could have been faster...), is that quassel is trying to look in qtbase.out/lib/qt-qtversion/sqldrivers.

Manually editing quasselcore's wrapper to have QT_PLUGIN_PATH pointing to qtbase.bin/lib/qtversion/plugins grants me a running quasselcore!

strace sort-of concurs, but it appears to be even dumber:

```strace
access("/nix/store/j1s2a86gb75l4mf29fvrdwbmgbbzi1b3-quassel-daemon-qt5-0.13.1/bin/sqldrivers/.", F_OK) = -1 ENOENT (No such file or directory)
```
(it then proceeds to die)

So, I tried to be smart and rebuild qt with the relocatable feature disabled, but I either failed to disable that correctly, or it was useless to do so. I'll leave another build running in case I messed something up with the conditionals, but I'll hopefully be sleeping when it's done.

**Comment by DIzFer** [comment]
Ah, new day, new ideas. I'm still making another test build, but could it be just dismissing the qtbase-qtpluginpath.patch needed in 5.12 is what's causing trouble? I assume it was removed for the 5.14 derivation because it failed to apply.

I'm studying up the relevant file to understand how to update the patch, and meanwhile I'm building 5.12 _without_ that patch to see if it behaves like 5.14 does now.

**Comment by DIzFer** [comment]
Aha! Once again I find myself celebrating a failure :D

quasselcore from nixos-unstable fails to find the database plugins with Qt5.12 once I removed the patch in question from Qt's derivation! strace looks very similar too!

5/8 components built for quassel+Qt5.14, having updated the patch to work in 5.14. Let's see how that goes

**Comment by petabyteboy** [comment]
That's great, thanks for doing the research! I must have thought that this patch was not needed anymore for some reason, or maybe I just overlooked it. 
Can you push your result to this PR or to a branch in your nixpkgs fork?

**Comment by DIzFer** [comment]
SUCCESS

Nah, no worries. If a patch fails to apply on source update (as this one did), my first reflex would also be to drop it (what could go wrong anyways :D)
I'm going to do a few more manual checks and I'll push my changes. BRB

**Comment by DIzFer** [comment]
https://github.com/DIzFer/nixpkgs/commit/eae6e824c70cd619321adaba16ede1f2d5505df6

_please don't be a "works on my machine"_

**Comment by petabyteboy** [comment]
> [DIzFer@eae6e82](https://github.com/DIzFer/nixpkgs/commit/eae6e824c70cd619321adaba16ede1f2d5505df6)
> 
> _please don't be a "works on my machine"_

Nope, I can confirm this fixes QuasselCore :+1: 

screenshot

**Comment by petabyteboy** [comment]
The latest Quassel release does not build with Qt 5.14.x. There are patches for this, but they can not be easily backported. Appearantly the upstream Quassel maintainers have been very unresponsive since a long time, some of the main contributors are very frustrated about this and asked the maintainers if they could add someone to the team, but they refused. So it is uncertain when we will see a new release and I was thinking if we could just package the current master branch as "unstable-2020xxxx". There is also the possibility of a fork emerging from Quassel, in which case we would have to decide which version we package. What do you think about this @Phreedom @ttuegel ?

**Comment by petabyteboy** [comment]
I had a chat with one of the active Quassel contributors and asked them about the master branch. They advised against using it, because there have been breakages on the master branch recently. The release of Quassel 0.14 and 0.13.2 was confirmed for this month by a maintainer on IRC. I found out that most of the Qt5.14-related changes are deprecation warnings, so I tried again to get a working patch and succeeded. I was also able to fix the Falkon build by applying one small patch.

**Comment by romildo** [comment]
I have made a separate [PR for updating and fixing Lumina](https://github.com/NixOS/nixpkgs/pull/88675), as it is independent from these changes.

I have also checked that it works with the new version of Qt proposed in this patch.

Hence the three commits related to Lumina can be dropped from here.

**Comment by ttuegel** [comment]
> There is also the possibility of a fork emerging from Quassel, in which case we would have to decide which version we package.

Let's keep Quassel on Qt 5.12 and see what happens upstream. Qt 5.12 is an LTS release, so we don't need to drop it soon.

**Comment by petabyteboy** [comment]
Quassel upstream has not made a new release last month as they had announced, but I got a confirmation from them that the patch I added is okay and this version can be delivered to users.

**Comment by DIzFer** [comment]
I've been using this for a few weeks now. Although my amount of dependent software is limited to sddm, quasselcore, qutebrowser and krunner, none seem to have behaved any differently with 5.14. Hopefully that spells good news for 5.15 now :P

**Comment by ttuegel** [comment]
I pushed commits which update and fix digikam and mark libqtav broken.

**Comment by gebner** [comment]
I've pushed a commit to update `pyside2` to 5.14.  FreeCAD works now.

**Comment by ttuegel** [comment]
The commit I just pushed for `qttools` fixes `diffpdf`.

**Comment by petabyteboy** [comment]
Thank you so much for fixing these remaining issues. I think we are getting very close to mergable. I will have a look at the remaining todos later.

**Review by jraygauthier — COMMENTED** [review]
Looked mostly the digikam changes. Built digikam from the PR successfully. Most things seem to work fine. However, missing video thumbnails and previews.

**Comment by nrdxp** [comment]
I went ahead and applied this work to 5.15 for the sake of using an updated qutebrowser, I haven't tested anything else so for now here is the [draft pr](https://github.com/petabyteboy/nixpkgs/pull/1).

**Comment by petabyteboy** [comment]
I tested some more applications and only had to change one more application (sdrangel) to Qt 5.12.

The deepin packages (specifically deepin.dde-api) don't build on master, so I couldn't check if this breaks anything. If there are no active maintainers I think we should probably put a eval warning in place to inform users that their package needs some help.
(CC @romildo)

netease-cloud-music builds when it and the deepin packages, and thus qcef, are changed to be built with Qt 5.12. However both on master and with this fix, it crashes on startup, so I did not make an effort to commit this change.
(CC @Ninlives)

I think this is finally getting ready to be merged. I think we should first get this through and then tackle Qt 5.15 in a seperate PR.

**Comment by worldofpeace** [comment]
Fabulous work here :+1:

**Comment by marcus7070** [comment]
You would expect `python3Packages.pyqtwebengine` to build from this PR, right? I can't get it to build.

**Comment by gebner** [comment]
> You would expect `python3Packages.pyqtwebengine` to build from this PR, right? I can't get it to build.

There are two python bindings for Qt, `pyside2` and `pyqt5`.  I've only fixed `pyside2` so far.

**Comment by romildo** [comment]
> The deepin packages (specifically deepin.dde-api) don't build on master, so I couldn't check if this breaks anything. If there are no active maintainers I think we should probably put a eval warning in place to inform users that their package needs some help.
> (CC @romildo)

I have been updating the deepin packages to current releases (see the draft PR https://github.com/NixOS/nixpkgs/pull/92625), but they have shown buggy. Also packaging deepin has shown very difficult because it relies very much on FHS compliant files that should be fixed for NixOS.

**Comment by colemickens** [comment]
I've been running this as my system config for a while. Works well. Thank you for doing this work!

**Comment by reinhardt** [comment]
I'm getting an error when installing qutebrowser with this branch:

`/build/PyQtWebEngine-5.15.0/QtWebEngineWidgets/sipQtWebEngineWidgetsQWebEngineCertificateError.cpp:368:99: error: ‘CertificateKnownInterceptionBlocked’ is not a member of ‘QWebEngineCertificateError’`

Is it worth writing a full report of this?

**Comment by gebner** [comment]
@reinhardt @marcus7070 Apparently `pyqtwebengine` (and `qutebrowser`) is also broken on master: #92500 :scream: 

I think we should merge this PR now, and then get `pyqtwebengine` working after that.  I'm tired of building qt locally.  We probably also need to revert #92500, because I don't expect the 5.15 version of `pyqtwebengine` to work with 5.14.

**Comment by petabyteboy** [comment]
That's very interesting. I think if we can confirm that PyQtWebEngine 5.15 only works with Qt 5.15 we might want to revert that PyQtWebEngine update until we have Qt 5.15. I successfully built a qutebrowser with @nrdxp's Qt 5.15 browser and PyQtWebEngine 5.15 and I'm using that right now:


```
[pbb@onyx:~]$ qutebrowser -V
         ______     ,,
    ,.-"`      | ,-` |
  .^           ||    |
 /    ,-*^|    ||    |
;    /    |    ||    ;-*```^*.
;   ;     |    |;,-*`         \
|   |     |  ,-*`    ,-"""\    \
|    \   ,-"`    ,-^`|     \    |
 \    `^^    ,-;|    |     ;    |
  *;     ,-*`  ||    |     /   ;;
    `^^`` |    ||    |   ,^    /
          |    ||    `^^`    ,^
          |  _,"|        _,-"
          -*`   ****"""``

qutebrowser v1.13.1
Git commit:
Backend: QtWebEngine (Chromium 80.0.3987.163)
Qt: 5.15.0

CPython: 3.8.3
PyQt: 5.15.0

sip: 4.19.23
colorama: no
pypeg2: 2.15
jinja2: 2.11.2
pygments: 2.6.1
yaml: 5.3.1
cssutils: 1.0.2 $Id$
attr: 19.3.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.0
PyQt5.QtWebKitWidgets: no
pdf.js: 2.3.200 (/nix/store/fvdbc4d34dsc94n2rpfhld8x452ryb6y-pdfjs-2.3.200/build/pdf.js)
sqlite: 3.32.3
QtNetwork SSL: OpenSSL 1.1.1g  21 Apr 2020

Style: QFusionStyle
Platform plugin: wayland
OpenGL: X.Org, 4.6 (Compatibility Profile) Mesa 20.1.3
Platform: Linux-5.8.0-rc6-x86_64-with-glibc2.2.5, 64bit
Linux distribution: NixOS 20.09 (Nightingale) (unknown)
Frozen: False
Imported from /nix/store/ac5cvzdrr2n9jbgn834psf79lgd5689n-qutebrowser-1.13.1/lib/python3.8/site-packages/qutebrowser
Using Python from /nix/store/8hf2rpqx8fnx6a3jm9pvn0d40c065zxl-python3-3.8.3/bin/python3.8
Qt library executable path: /nix/store/6sq8kb88ybf95y8b6kdc5kpzdpiyr0iz-qtbase-5.15.0/libexec, data path: /nix/store/6sq8kb88ybf95y8b6kdc5kpzdpiyr0iz-qtbase-5.15.0
OS Version:

--- /etc/os-release ---
NAME=NixOS
ID=nixos
VERSION="20.09.git.a8546a078dfM (Nightingale)"
VERSION_CODENAME=nightingale
VERSION_ID="20.09.git.a8546a078dfM"
PRETTY_NAME="NixOS 20.09 (Nightingale)"
LOGO="nix-snowflake"
DOCUMENTATION_URL="https://nixos.org/learn.html"

Paths:
cache: /home/pbb/.cache/qutebrowser
config: /home/pbb/.config/qutebrowser
data: /home/pbb/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser
system data: /nix/store/ac5cvzdrr2n9jbgn834psf79lgd5689n-qutebrowser-1.13.1/share/qutebrowser

Autoconfig loaded: no
Config.py: /home/pbb/.config/qutebrowser/config.py has been loaded
Uptime: 0:00:00

[pbb@onyx:~]$
```

**Comment by marcus7070** [comment]
@gebner, that's weird, I get a successful build from f52e3212cb43eabc59a2fe5b026ebd7869411604. And I'm pretty sure I tested the commit where this PR diverged from master and that built too. I definetly got a successful build with pyqtwebengine v5.15 (and PyQt5.12) yesterday. 

Hydra also suggests master is good: https://hydra.nixos.org/eval/1603145?filter=pyqtwebengine&compare=1603133&full=#tabs-still-succeed

Just to be clear I'm not arguing this shouldn't or should be merged as is, just trying to help.

**Comment by petabyteboy** [comment]
Thanks @marcus7070, that's good to know, so the failure must be caused by something different than the combination of Qt <5.15 and PyQtWebEngine 5.15. 
@reinhart did you build from this branch without any rebase or change? So commit 8fdd19ffa4029ff7418dfb89c390b6b70479b48e?

**Review by worldofpeace — APPROVED** [review]
:shipit: LGTM

**Comment by worldofpeace** [comment]
This should go to staging, though.

**Comment by petabyteboy** [comment]
I can reproduce the PyQtWebEngine build failure and I'm now trying with a rebased version, but I have to build qtwebengine first, so it could take an hour or so...

**Comment by The-Compiler** [comment]
FWIW PyQtWebEngine (and PyQt in general) can run fine against a different version of Qt (except when it accidentally doesn't build, like [it was the case before](https://www.riverbankcomputing.com/pipermail/pyqt/2020-June/042985.html)). The upstream recommendation is to always use the newest PyQt, even against older Qt versions.

**Comment by petabyteboy** [comment]
This is getting really confusing :smile:
Qt 5.12.x and QtWebEngine 5.15 => Works
Qt 5.14.x and QtWebEngine 5.15 => Doesn't work
Qt 5.15.x and QtWebEngine 5.15 => Works
Thanks @The-Compiler for pointing us to that mailing list entry. I will add the patch you posted there.

**Comment by reinhardt** [comment]
> @Reinhart did you build from this branch without any rebase or change? So commit [8fdd19f](https://github.com/NixOS/nixpkgs/commit/8fdd19ffa4029ff7418dfb89c390b6b70479b48e)?

Yup, exactly. That was the first error I got, before touching anything. Later I fiddled with the versions a bit and got it to build but had trouble with ssl at runtime.

**Comment by petabyteboy** [comment]
Unless anyone raises new or existing concerns I will merge this today

**Review by FRidh — APPROVED** [review]

**Comment by jonringer** [comment]
nice work everyone :)

**Comment by vcunat** [comment]
Can some of you have a look at the channel-blocking darwin failure?  https://github.com/NixOS/nixpkgs/issues/95199

**Pull request merged**
